File: platforms/ios/Dezel/DezelTests/Core/JavaScriptPropertyTest.swift
import XCTest
import Dezel

class JavaScriptPropertyTest: XCTestCase {

	private var context: JavaScriptContext = JavaScriptContext()

	override func setUp() {
		super.setUp()
		self.context.dispose()
		self.context = JavaScriptContext()
	}

	func testPropertyCreatedWithNull() {

		let property = JavaScriptProperty()

		XCTAssertEqual(property.type, .null)
		XCTAssertEqual(property.unit, .none)

		XCTAssertEqual(property.isNull, true)
		XCTAssertEqual(property.isString, false)
		XCTAssertEqual(property.isNumber, false)
		XCTAssertEqual(property.isBoolean, false)
		XCTAssertEqual(property.isObject, false)
		XCTAssertEqual(property.isArray, false)

		XCTAssertEqual(property.string, "")
		XCTAssertEqual(property.number, 0.0)
		XCTAssertEqual(property.boolean, false)
	}

	func testPropertyCreateWithString() {

		let property = JavaScriptProperty(string: "test")

		XCTAssertEqual(property.type, .string)
		XCTAssertEqual(property.unit, .none)

		XCTAssertEqual(property.isNull, false)
		XCTAssertEqual(property.isString, true)
		XCTAssertEqual(property.isNumber, false)
		XCTAssertEqual(property.isBoolean, false)
		XCTAssertEqual(property.isObject, false)
		XCTAssertEqual(property.isArray, false)

		XCTAssertEqual(property.string, "test")
		XCTAssertEqual(property.number.isNaN, true)
		XCTAssertEqual(property.boolean, true)
	}

	func testPropertyCreatedWithNumber() {

		let property = JavaScriptProperty(number: 14.0)

		XCTAssertEqual(property.type, .number)
		XCTAssertEqual(property.unit, .none)

		XCTAssertEqual(property.isNull, false)
		XCTAssertEqual(property.isString, false)
		XCTAssertEqual(property.isNumber, true)
		XCTAssertEqual(property.isBoolean, false)
		XCTAssertEqual(property.isObject, false)
		XCTAssertEqual(property.isArray, false)

		XCTAssertEqual(property.string, "14")
		XCTAssertEqual(property.number, 14.0)
		XCTAssertEqual(property.boolean, true)
	}

	func testPropertyCreatedWithNumberAndUnit() {

		let property = JavaScriptProperty(number: 14.0, unit: .px)

		XCTAssertEqual(property.type, .number)
		XCTAssertEqual(property.unit, .px)

		XCTAssertEqual(property.isNull, false)
		XCTAssertEqual(property.isString, false)
		XCTAssertEqual(property.isNumber, true)
		XCTAssertEqual(property.isBoolean, false)
		XCTAssertEqual(property.isObject, false)
		XCTAssertEqual(property.isArray, false)

		XCTAssertEqual(property.string, "14px")
		XCTAssertEqual(property.number, 14.0)
		XCTAssertEqual(property.boolean, true)
	}

	func testPropertyCreatedWithBoolean() {

		let property = JavaScriptProperty(boolean: true)

		XCTAssertEqual(property.type, .boolean)
		XCTAssertEqual(property.unit, .none)

		XCTAssertEqual(property.isNull, false)
		XCTAssertEqual(property.isString, false)
		XCTAssertEqual(property.isNumber, false)
		XCTAssertEqual(property.isBoolean, true)
		XCTAssertEqual(property.isObject, false)
		XCTAssertEqual(property.isArray, false)

		XCTAssertEqual(property.string, "true")
		XCTAssertEqual(property.number, 1.0)
		XCTAssertEqual(property.boolean, true)
	}

	func testPropertyStringConversion() {

		let n1 = JavaScriptProperty(number: 5.0)
		let n2 = JavaScriptProperty(number: 5.5)
		let n3 = JavaScriptProperty(number: -5.0)
		let n4 = JavaScriptProperty(number: -5.5)

		XCTAssertEqual(n1.string, "5")
		XCTAssertEqual(n2.string, "5.5")
		XCTAssertEqual(n3.string, "-5")
		XCTAssertEqual(n4.string, "-5.5")

		let npx = JavaScriptProperty(number: 5.0, unit: .px)
		let npc = JavaScriptProperty(number: 5.0, unit: .pc)
		let nvw = JavaScriptProperty(number: 5.0, unit: .vw)
		let nvh = JavaScriptProperty(number: 5.0, unit: .vh)
		let npw = JavaScriptProperty(number: 5.0, unit: .pw)
		let nph = JavaScriptProperty(number: 5.0, unit: .ph)
		let ncw = JavaScriptProperty(number: 5.0, unit: .cw)
		let nch = JavaScriptProperty(number: 5.0, unit: .ch)

		XCTAssertEqual(npx.string, "5px")
		XCTAssertEqual(npc.string, "5%")
		XCTAssertEqual(nvw.string, "5vw")
		XCTAssertEqual(nvh.string, "5vh")
		XCTAssertEqual(npw.string, "5pw")
		XCTAssertEqual(nph.string, "5ph")
		XCTAssertEqual(ncw.string, "5cw")
		XCTAssertEqual(nch.string, "5ch")

		let b1 = JavaScriptProperty(boolean: true)
		let b2 = JavaScriptProperty(boolean: false)

		XCTAssertEqual(b1.string, "true")
		XCTAssertEqual(b2.string, "false")
	}

	func testPropertyNumberConversion() {

		let s1 = JavaScriptProperty(string: "5.0")
		let s2 = JavaScriptProperty(string: "5.5")
		let s3 = JavaScriptProperty(string: "-5.0")
		let s4 = JavaScriptProperty(string: "-5.5")

		XCTAssertEqual(s1.number, 5)
		XCTAssertEqual(s2.number, 5.5)
		XCTAssertEqual(s3.number, -5)
		XCTAssertEqual(s4.number, -5.5)

		let b1 = JavaScriptProperty(boolean: true)
		let b2 = JavaScriptProperty(boolean: false)

		XCTAssertEqual(b1.number, 1.0)
		XCTAssertEqual(b2.number, 0.0)
	}

	func testPropertyBooleanConversion() {

		let s1 = JavaScriptProperty(string: "5.0")
		let s2 = JavaScriptProperty(string: "0.0")
		let s3 = JavaScriptProperty(string: "")
		let s4 = JavaScriptProperty(string: "value")

		XCTAssertEqual(s1.boolean, true)
		XCTAssertEqual(s2.boolean, true)
		XCTAssertEqual(s3.boolean, false)
		XCTAssertEqual(s4.boolean, true)

		let b1 = JavaScriptProperty(number: 5.0)
		let b2 = JavaScriptProperty(number: 0.0)

		XCTAssertEqual(b1.boolean, true)
		XCTAssertEqual(b2.boolean, false)
	}

	func testResetWithJavaScriptStringValue() {

		let value = self.context.createString("test")

		let property = JavaScriptProperty()

		property.reset(value)

		XCTAssertEqual(property.type, .string)
		XCTAssertEqual(property.unit, .none)

		XCTAssertEqual(property.string, "test")
		XCTAssertEqual(property.number.isNaN, true)
		XCTAssertEqual(property.boolean, true)

		XCTAssertEqual(property.toHandle(self.context), value.handle)
	}

	func testResetWithJavaScriptStringWithUnitValue() {

		let px = self.context.createString("15px")
		let pc = self.context.createString("15%")
		let vw = self.context.createString("15vw")
		let vh = self.context.createString("15vh")
		let pw = self.context.createString("15pw")
		let ph = self.context.createString("15ph")
		let cw = self.context.createString("15cw")
		let ch = self.context.createString("15ch")

		let property = JavaScriptProperty()

		property.reset(px, parse: true)

		XCTAssertEqual(property.type, .number)
		XCTAssertEqual(property.unit, .px)

		XCTAssertEqual(property.string, "15px")
		XCTAssertEqual(property.number, 15)
		XCTAssertEqual(property.boolean, true)

		XCTAssertEqual(property.toHandle(self.context), px.handle)

		property.reset(pc, parse: true)

		XCTAssertEqual(property.type, .number)
		XCTAssertEqual(property.unit, .pc)

		XCTAssertEqual(property.string, "15%")
		XCTAssertEqual(property.number, 15)
		XCTAssertEqual(property.boolean, true)

		XCTAssertEqual(property.toHandle(self.context), pc.handle)

		property.reset(vw, parse: true)

		XCTAssertEqual(property.type, .number)
		XCTAssertEqual(property.unit, .vw)

		XCTAssertEqual(property.string, "15vw")
		XCTAssertEqual(property.number, 15)
		XCTAssertEqual(property.boolean, true)

		XCTAssertEqual(property.toHandle(self.context), vw.handle)

		property.reset(vh, parse: true)

		XCTAssertEqual(property.type, .number)
		XCTAssertEqual(property.unit, .vh)

		XCTAssertEqual(property.string, "15vh")
		XCTAssertEqual(property.number, 15)
		XCTAssertEqual(property.boolean, true)

		XCTAssertEqual(property.toHandle(self.context), vh.handle)

		property.reset(pw, parse: true)

		XCTAssertEqual(property.type, .number)
		XCTAssertEqual(property.unit, .pw)

		XCTAssertEqual(property.string, "15pw")
		XCTAssertEqual(property.number, 15)
		XCTAssertEqual(property.boolean, true)

		XCTAssertEqual(property.toHandle(self.context), pw.handle)

		property.reset(ph, parse: true)

		XCTAssertEqual(property.type, .number)
		XCTAssertEqual(property.unit, .ph)

		XCTAssertEqual(property.string, "15ph")
		XCTAssertEqual(property.number, 15)
		XCTAssertEqual(property.boolean, true)

		XCTAssertEqual(property.toHandle(self.context), ph.handle)

		property.reset(cw, parse: true)

		XCTAssertEqual(property.type, .number)
		XCTAssertEqual(property.unit, .cw)

		XCTAssertEqual(property.string, "15cw")
		XCTAssertEqual(property.number, 15)
		XCTAssertEqual(property.boolean, true)

		XCTAssertEqual(property.toHandle(self.context), cw.handle)

		property.reset(ch, parse: true)

		XCTAssertEqual(property.type, .number)
		XCTAssertEqual(property.unit, .ch)

		XCTAssertEqual(property.string, "15ch")
		XCTAssertEqual(property.number, 15)
		XCTAssertEqual(property.boolean, true)

		XCTAssertEqual(property.toHandle(self.context), ch.handle)
	}

	func testResetWithJavaScriptNumberValue() {

		let value = self.context.createNumber(15.0)

		let property = JavaScriptProperty()

		property.reset(value)

		XCTAssertEqual(property.type, .number)
		XCTAssertEqual(property.unit, .none)

		XCTAssertEqual(property.string, "15")
		XCTAssertEqual(property.number, 15)
		XCTAssertEqual(property.boolean, true)

		XCTAssertEqual(property.toHandle(self.context), value.handle)
	}

	func testResetWithJavaScriptBooleanValue() {

		let value = self.context.createBoolean(true)

		let property = JavaScriptProperty()

		property.reset(value)

		XCTAssertEqual(property.type, .boolean)
		XCTAssertEqual(property.unit, .none)

		XCTAssertEqual(property.string, "true")
		XCTAssertEqual(property.number, 1)
		XCTAssertEqual(property.boolean, true)

		XCTAssertEqual(property.toHandle(self.context), value.handle)
	}

	func testChangeCallback() {

		let expectation = self.expectation(description: "Called")

		let property = JavaScriptProperty() { value in
			expectation.fulfill()
		}

		property.reset("test")

		self.wait(for: [expectation], timeout: 1.0)
	}

	func testChangeCallbackWithTypes() {

		var called = 0

		let property = JavaScriptProperty() { value in
			called += 1
		}

		property.reset("test")
		property.reset("test")
		property.reset(10.0)
		property.reset(10.0, unit: .px)
		property.reset(10.0, unit: .px)
		property.reset(true)
		property.reset(true)

		XCTAssertEqual(called, 4)
	}
}

File: platforms/android/Dezel/dezel/src/androidTest/java/ca/logaritm/dezel/core/JavaScriptPropertyTest.kt
package ca.logaritm.dezel.core

import android.util.Log
import org.junit.Assert
import org.junit.Before
import org.junit.Test

class JavaScriptPropertyTest {

	private var context: JavaScriptContext = JavaScriptContext()

	@Before
	fun beforeTest() {
		this.context.dispose()
		this.context = JavaScriptContext()
	}
	
	@Test
	fun testPropertyCreatedWithNull() {

		val property = JavaScriptProperty()

		Assert.assertEquals(property.type, JavaScriptPropertyType.NULL)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.NONE)

		Assert.assertEquals(property.isNull, true)
		Assert.assertEquals(property.isString, false)
		Assert.assertEquals(property.isNumber, false)
		Assert.assertEquals(property.isBoolean, false)
		Assert.assertEquals(property.isObject, false)
		Assert.assertEquals(property.isArray, false)

		Assert.assertEquals(property.string, "")
		Assert.assertEquals(property.number, 0.0, 0.0)
		Assert.assertEquals(property.boolean, false)
	}

	@Test
	fun testPropertyCreateWithString() {

		val property = JavaScriptProperty("test")

		Assert.assertEquals(property.type, JavaScriptPropertyType.STRING)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.NONE)

		Assert.assertEquals(property.isNull, false)
		Assert.assertEquals(property.isString, true)
		Assert.assertEquals(property.isNumber, false)
		Assert.assertEquals(property.isBoolean, false)
		Assert.assertEquals(property.isObject, false)
		Assert.assertEquals(property.isArray, false)

		Assert.assertEquals(property.string, "test")
		Assert.assertEquals(property.number.isNaN(), true)
		Assert.assertEquals(property.boolean, true)
	}

	@Test
	fun testPropertyCreatedWithNumber() {

		val property = JavaScriptProperty(14.0)

		Assert.assertEquals(property.type, JavaScriptPropertyType.NUMBER)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.NONE)

		Assert.assertEquals(property.isNull, false)
		Assert.assertEquals(property.isString, false)
		Assert.assertEquals(property.isNumber, true)
		Assert.assertEquals(property.isBoolean, false)
		Assert.assertEquals(property.isObject, false)
		Assert.assertEquals(property.isArray, false)

		Assert.assertEquals(property.string, "14")
		Assert.assertEquals(property.number, 14.0, 0.0)
		Assert.assertEquals(property.boolean, true)
	}

	@Test
	fun testPropertyCreatedWithNumberAndUnit() {

		val property = JavaScriptProperty(14.0, JavaScriptPropertyUnit.PX)

		Assert.assertEquals(property.type, JavaScriptPropertyType.NUMBER)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.PX)

		Assert.assertEquals(property.isNull, false)
		Assert.assertEquals(property.isString, false)
		Assert.assertEquals(property.isNumber, true)
		Assert.assertEquals(property.isBoolean, false)
		Assert.assertEquals(property.isObject, false)
		Assert.assertEquals(property.isArray, false)

		Assert.assertEquals(property.string, "14px")
		Assert.assertEquals(property.number, 14.0, 0.0)
		Assert.assertEquals(property.boolean, true)
	}

	@Test
	fun testPropertyCreatedWithBoolean() {

		val property = JavaScriptProperty(true)

		Assert.assertEquals(property.type, JavaScriptPropertyType.BOOLEAN)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.NONE)

		Assert.assertEquals(property.isNull, false)
		Assert.assertEquals(property.isString, false)
		Assert.assertEquals(property.isNumber, false)
		Assert.assertEquals(property.isBoolean, true)
		Assert.assertEquals(property.isObject, false)
		Assert.assertEquals(property.isArray, false)

		Assert.assertEquals(property.string, "true")
		Assert.assertEquals(property.number, 1.0, 0.0)
		Assert.assertEquals(property.boolean, true)
	}

	@Test
	fun testPropertyStringConversion() {

		val n1 = JavaScriptProperty(5.0)
		val n2 = JavaScriptProperty(5.5)
		val n3 = JavaScriptProperty(-5.0)
		val n4 = JavaScriptProperty(-5.5)

		Assert.assertEquals(n1.string, "5")
		Assert.assertEquals(n2.string, "5.5")
		Assert.assertEquals(n3.string, "-5")
		Assert.assertEquals(n4.string, "-5.5")

		val npx = JavaScriptProperty(5.0, JavaScriptPropertyUnit.PX)
		val npc = JavaScriptProperty(5.0, JavaScriptPropertyUnit.PC)
		val nvw = JavaScriptProperty(5.0, JavaScriptPropertyUnit.VW)
		val nvh = JavaScriptProperty(5.0, JavaScriptPropertyUnit.VH)
		val npw = JavaScriptProperty(5.0, JavaScriptPropertyUnit.PW)
		val nph = JavaScriptProperty(5.0, JavaScriptPropertyUnit.PH)
		val ncw = JavaScriptProperty(5.0, JavaScriptPropertyUnit.CW)
		val nch = JavaScriptProperty(5.0, JavaScriptPropertyUnit.CH)

		Assert.assertEquals(npx.string, "5px")
		Assert.assertEquals(npc.string, "5%")
		Assert.assertEquals(nvw.string, "5vw")
		Assert.assertEquals(nvh.string, "5vh")
		Assert.assertEquals(npw.string, "5pw")
		Assert.assertEquals(nph.string, "5ph")
		Assert.assertEquals(ncw.string, "5cw")
		Assert.assertEquals(nch.string, "5ch")

		val b1 = JavaScriptProperty(true)
		val b2 = JavaScriptProperty(false)

		Assert.assertEquals(b1.string, "true")
		Assert.assertEquals(b2.string, "false")
	}

	@Test
	fun testPropertyNumberConversion() {

		val s1 = JavaScriptProperty("5.0")
		val s2 = JavaScriptProperty("5.5")
		val s3 = JavaScriptProperty("-5.0")
		val s4 = JavaScriptProperty("-5.5")

		Assert.assertEquals(s1.number, 5.0, 0.0)
		Assert.assertEquals(s2.number, 5.5, 0.0)
		Assert.assertEquals(s3.number, -5.0, 0.0)
		Assert.assertEquals(s4.number, -5.5, 0.0)

		val b1 = JavaScriptProperty(true)
		val b2 = JavaScriptProperty(false)

		Assert.assertEquals(b1.number, 1.0, 0.0)
		Assert.assertEquals(b2.number, 0.0, 0.0)
	}

	@Test
	fun testPropertyBooleanConversion() {

		val s1 = JavaScriptProperty("5.0")
		val s2 = JavaScriptProperty("0.0")
		val s3 = JavaScriptProperty("")
		val s4 = JavaScriptProperty("value")

		Assert.assertEquals(s1.boolean, true)
		Assert.assertEquals(s2.boolean, true)
		Assert.assertEquals(s3.boolean, false)
		Assert.assertEquals(s4.boolean, true)

		val b1 = JavaScriptProperty(5.0)
		val b2 = JavaScriptProperty(0.0)

		Assert.assertEquals(b1.boolean, true)
		Assert.assertEquals(b2.boolean, false)
	}

	@Test
	fun testResetWithJavaScriptStringValue() {

		val value = this.context.createString("test")

		val property = JavaScriptProperty()

		property.reset(value)

		Assert.assertEquals(property.type, JavaScriptPropertyType.STRING)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.NONE)

		Assert.assertEquals(property.string, "test")
		Assert.assertEquals(property.number.isNaN(), true)
		Assert.assertEquals(property.boolean, true)

		Assert.assertEquals(property.toHandle(this.context), value.handle)
	}

	@Test
	fun testResetWithJavaScriptStringWithUnitValue() {

		val px = this.context.createString("15px")
		val pc = this.context.createString("15%")
		val vw = this.context.createString("15vw")
		val vh = this.context.createString("15vh")
		val pw = this.context.createString("15pw")
		val ph = this.context.createString("15ph")
		val cw = this.context.createString("15cw")
		val ch = this.context.createString("15ch")

		val property = JavaScriptProperty()

		property.reset(px)

		Assert.assertEquals(property.type, JavaScriptPropertyType.NUMBER)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.PX)

		Assert.assertEquals(property.string, "15px")
		Assert.assertEquals(property.number, 15.0, 0.0)
		Assert.assertEquals(property.boolean, true)

		Assert.assertEquals(property.toHandle(this.context), px.handle)

		property.reset(pc)

		Assert.assertEquals(property.type, JavaScriptPropertyType.NUMBER)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.PC)

		Assert.assertEquals(property.string, "15%")
		Assert.assertEquals(property.number, 15.0, 0.0)
		Assert.assertEquals(property.boolean, true)

		Assert.assertEquals(property.toHandle(this.context), pc.handle)

		property.reset(vw)

		Assert.assertEquals(property.type, JavaScriptPropertyType.NUMBER)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.VW)

		Assert.assertEquals(property.string, "15vw")
		Assert.assertEquals(property.number, 15.0, 0.0)
		Assert.assertEquals(property.boolean, true)

		Assert.assertEquals(property.toHandle(this.context), vw.handle)

		property.reset(vh)

		Assert.assertEquals(property.type, JavaScriptPropertyType.NUMBER)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.VH)

		Assert.assertEquals(property.string, "15vh")
		Assert.assertEquals(property.number, 15.0, 0.0)
		Assert.assertEquals(property.boolean, true)

		Assert.assertEquals(property.toHandle(this.context), vh.handle)

		property.reset(pw)

		Assert.assertEquals(property.type, JavaScriptPropertyType.NUMBER)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.PW)

		Assert.assertEquals(property.string, "15pw")
		Assert.assertEquals(property.number, 15.0, 0.0)
		Assert.assertEquals(property.boolean, true)

		Assert.assertEquals(property.toHandle(this.context), pw.handle)

		property.reset(ph)

		Assert.assertEquals(property.type, JavaScriptPropertyType.NUMBER)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.PH)

		Assert.assertEquals(property.string, "15ph")
		Assert.assertEquals(property.number, 15.0, 0.0)
		Assert.assertEquals(property.boolean, true)

		Assert.assertEquals(property.toHandle(this.context), ph.handle)

		property.reset(cw)

		Assert.assertEquals(property.type, JavaScriptPropertyType.NUMBER)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.CW)

		Assert.assertEquals(property.string, "15cw")
		Assert.assertEquals(property.number, 15.0, 0.0)
		Assert.assertEquals(property.boolean, true)

		Assert.assertEquals(property.toHandle(this.context), cw.handle)

		property.reset(ch)

		Assert.assertEquals(property.type, JavaScriptPropertyType.NUMBER)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.CH)

		Assert.assertEquals(property.string, "15ch")
		Assert.assertEquals(property.number, 15.0, 0.0)
		Assert.assertEquals(property.boolean, true)

		Assert.assertEquals(property.toHandle(this.context), ch.handle)
	}

	@Test
	fun testResetWithJavaScriptNumberValue() {

		val value = this.context.createNumber(15.0)

		val property = JavaScriptProperty()

		property.reset(value)

		Assert.assertEquals(property.type, JavaScriptPropertyType.NUMBER)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.NONE)

		Assert.assertEquals(property.string, "15")
		Assert.assertEquals(property.number, 15.0, 0.0)
		Assert.assertEquals(property.boolean, true)

		Assert.assertEquals(property.toHandle(this.context), value.handle)
	}

	@Test
	fun testResetWithJavaScriptBooleanValue() {

		val value = this.context.createBoolean(true)

		val property = JavaScriptProperty()

		property.reset(value)

		Assert.assertEquals(property.type, JavaScriptPropertyType.BOOLEAN)
		Assert.assertEquals(property.unit, JavaScriptPropertyUnit.NONE)

		Assert.assertEquals(property.string, "true")
		Assert.assertEquals(property.number, 1.0, 0.0)
		Assert.assertEquals(property.boolean, true)

		Assert.assertEquals(property.toHandle(this.context), value.handle)
	}

	@Test
	fun testChangeCallback() {

		var called = false

		val property = JavaScriptProperty() { value ->
			called = true
		}

		property.reset("test")

		Assert.assertTrue(called)
	}

	@Test
	fun testChangeCallbackWithTypes() {

		var called = 0

		val property = JavaScriptProperty() { value ->
			called += 1
		}

		property.reset("test")
		property.reset("test")
		property.reset(10.0)
		property.reset(10.0, JavaScriptPropertyUnit.PX)
		property.reset(10.0, JavaScriptPropertyUnit.PX)
		property.reset(true)
		property.reset(true)

		Assert.assertEquals(called, 4)
	}
}
